bluesky/callbacks/zmq.py

```diff
@@ -0,0 +1,238 @@
+            self._frontend = frontend
+            self._backend = backend
+            self._context = context
+        
+    def start(self):
+        if self.closed:
+            raise RuntimeError("This Proxy has already been started and "
+                               "interrupted. Create a fresh instance with "
+                               "{}".format(repr(self)))
+        try:
+            zmq.device(zmq.FORWARDER, self._frontend, self._backend)
+        finally:
+            self.closed = True
+            self._frontend.close()
+            self._backend.close()
+            self._context.term()
+
+    def __repr__(self):
+        return ("{}(in_port={in_port}, out_port={out_port})"
+                "".format(type(self).__name__, **vars(self)))
+
+
+class RemoteDispatcher(Dispatcher):
+    """
+    Dispatch documents received over the network from a 0MQ proxy.
+
+    Parameters
+    ----------
+    address : tuple
+        Address of a running 0MQ proxy, given either as a string like
+        ``'127.0.0.1:5567'`` or as a tuple like ``('127.0.0.1', 5567)``
+    hostname : string, optional
+        A filter: only process documents from a host with this name.
+    pid : int, optional
+        A filter: only process documents from a process with this pid.
+    run_engine_id : int, optional
+        A filter: only process documents from a RunEngine with this Python id
+        (memory address).
+    loop : zmq.asyncio.ZMQEventLoop, optional
+
+    Example
+    -------
+
+    Print all documents generated by remote RunEngines.
+
+    >>> d = RemoteDispatcher(('localhost', 5568))
+    >>> d.subscribe('all', print)
+    >>> d.start()  # runs until interrupted
+    """
+    def __init__(self, address, *, hostname=None, pid=None, run_engine_id=None,
+                 loop=None):
+        if isinstance(address, str):
+            address = address.split(':')
+        self.address = (address[0], int(address[1]))
+        self.hostname = hostname
+        self.pid = pid
+        self.run_engine_id = run_engine_id
+
+        if loop is None:
+            loop = zmq.asyncio.ZMQEventLoop()
+        self._loop = loop
+        asyncio.set_event_loop(self._loop)
+        self._context = zmq.asyncio.Context()
+        self._socket = self._context.socket(zmq.SUB)
+        url = "tcp://%s:%d" % self.address
+        self._socket.connect(url)
+        self._socket.setsockopt_string(zmq.SUBSCRIBE, "")
+        self._task = None
+
+        def is_our_message(_hostname, _pid, _RE_id):
+            # Close over filters and decide if this message applies to this
+            # RemoteDispatcher.
+            return ((hostname is None or hostname == _hostname)
+                    and (pid is None or pid == _pid)
+                    and (run_engine_id is None or run_engine_id ==
+                         run_engine_id))
+        self._is_our_message = is_our_message
+
+        super().__init__()
+
+    @asyncio.coroutine
+    def _poll(self):
```

Comment:
These lines are marked as not covered by Coverage because they are executed in a subprocess. I burned two hours trying to write a separate test running them in the main process, and gave up because of bad interactions between pytest and zmq.asyncio's event loop. We do test that these lines actually execute the callback registry, so I'm comfortable leaving it alone.